services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/factory/HakemusUpdateRequestFactory.kt

```diff
@@ -25,4 +37,79 @@ object HakemusUpdateRequestFactory {
             representativeWithContacts = null
         )
     }
+
+    fun createFilledJohtoselvityshakemusUpdateRequest(): JohtoselvityshakemusUpdateRequest {
+        return JohtoselvityshakemusUpdateRequest(
+            name = "Testihakemus",
+            postalAddress = PostalAddressRequest(StreetAddress("Kotikatu 1")),
+            constructionWork = true,
+            maintenanceWork = false,
+            propertyConnectivity = false,
+            emergencyWork = false,
+            rockExcavation = false,
+            workDescription = "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
+            startTime = ZonedDateTime.now(TZ_UTC),
+            endTime = ZonedDateTime.ofInstant(Instant.now().plus(5, ChronoUnit.DAYS), TZ_UTC),
+            areas = listOf(ApplicationArea("Hankealue 1", GeometriaFactory.polygon)),
+            customerWithContacts =
+                CustomerWithContactsRequest(
+                    // new customer without an id
+                    CustomerRequest(
+                        type = CustomerType.COMPANY,
+                        name = "Testiyritys",
+                        email = "info@testiyritys.fi",
+                        phone = "123456789",
+                        registryKey = "1234567-8",
+                    ),
+                    listOf(
+                        // an existing contact person
+                        ContactRequest(
+                            hankekayttajaId =
+                                UUID.fromString("cd1d4d2f-526b-4ee5-a1fa-97b14d25a11f")
+                        )
+                    )
+                ),
+            contractorWithContacts = null,
+            propertyDeveloperWithContacts = null,
+            representativeWithContacts = null
+        )
+    }
+
+    fun createJohtoselvityshakemusUpdateRequestFromApplicationEntity(
+        applicationEntity: ApplicationEntity
+    ): JohtoselvityshakemusUpdateRequest {
+        val applicationData = applicationEntity.applicationData as CableReportApplicationData
+        return JohtoselvityshakemusUpdateRequest(
+            name = applicationData.name,
+            postalAddress =
+                applicationData.postalAddress?.let { PostalAddressRequest(it.streetAddress) }
+                    ?: PostalAddressRequest(StreetAddress("")),
+            constructionWork = applicationData.constructionWork,
+            maintenanceWork = applicationData.maintenanceWork,
+            propertyConnectivity = applicationData.propertyConnectivity,
+            emergencyWork = applicationData.emergencyWork,
+            rockExcavation = applicationData.rockExcavation ?: false,
+            workDescription = applicationData.workDescription,
+            startTime = applicationData.startTime,
+            endTime = applicationData.endTime,
+            areas = applicationData.areas,
+        )
+    }
```

Comment:
If you get the HakemusResponse from HakemusService instead of using the entity directly, this can be simplified dramatically.

```suggestion
    fun HakemusResponse.toUpdateRequest(): JohtoselvityshakemusUpdateRequest =
        this.applicationData.toJsonString().parseJson()
```
This might seem like a hack at first glance, but this is exactly what the front end will do when they request an update. FE will read the DTO json, modify it and send the result as an update. So the response and update request need to be JSON-compatible in this direction. If this fails, it's a sign that the FE will fail here as well.

Reply:
I do not use HakemusService to create an Entity atm. I use HakemusFactory for that.

Reply:
But, you're right that the JSON-compatibility is an issue that I need to solve when I work on the Controller. I guess it needs similar @JsonTypeInfo and @JsonSubTypes annotations as are in ApplicationData class. And they need applicationType to be added which I was hoping to awoid since it is not necessary in business-sense. The type is already in the parent Application and copying it to ApplicationData seemed unnecessary. Until I noticed those JSON-annotations.

Reply:
> I do not use HakemusService to create an Entity atm. I use HakemusFactory for that.
You can call HakemusService after the entity has been saved.
```
            val entity = hakemusFactory.builder(USERID).save()
            val hakemus = hakemusService.hakemusResponse(entity.id!!)
```

Reply:
Yes, copying the application type to application data didn't feel like a perfect solution, but I couldn't think of any way around it. We could also solve the JSON subtyping when we start working on kaivuilmoitus, so we have a second type of application.

Reply:
I refactored the creation of hakemus and request as you suggested.